cirq-core/cirq/sim/classical_simulator.py

```diff
@@ -0,0 +1,110 @@
+          results. Measurement results are stored in a 3-dimensional
+          numpy array, the first dimension corresponding to the repetition.
+          the second to the instance of that key in the circuit, and the
+          third to the actual boolean measurement results (ordered by the
+          qubits being measured.)
+
+     Raises:
+      ValuesError: gate is not a cirq.XGate or cirq.Cnot
+    """
+
+    def _run(
+        self, circuit: AbstractCircuit, param_resolver: ParamResolver, repetitions: int
+    ) -> Dict[str, np.ndarray]:
+        results_dict: Dict[str, np.ndarray] = {}
+        values_dict: Dict[Qid, int] = defaultdict(int)
+        param_resolver = param_resolver or ParamResolver({})
+        resolved_circuit = protocols.resolve_parameters(circuit, param_resolver)
+
+        for moment in resolved_circuit:
+            for op in moment:
+                gate = op.gate
+                if isinstance(gate, ops.XPowGate) and gate.exponent == 1:
+                    values_dict[op.qubits[0]] = 1 - values_dict[op.qubits[0]]
+
+                elif isinstance(gate, ops.CNotPowGate) and gate.exponent == 1:
+                    if values_dict[op.qubits[0]] == 1:
+                        values_dict[op.qubits[1]] = 1 - values_dict[op.qubits[1]]
+
+                elif isinstance(gate, ops.SwapPowGate) and gate.exponent == 1:
+                    hold_qubit = values_dict[op.qubits[1]]
+                    values_dict[op.qubits[1]] = values_dict[op.qubits[0]]
+                    values_dict[op.qubits[0]] = hold_qubit
+
+                elif isinstance(gate, ops.CCXPowGate) and gate.exponent == 1:
+                    if (values_dict[op.qubits[0]] == 1) and (values_dict[op.qubits[1]] == 1):
+                        values_dict[op.qubits[2]] = 1 - values_dict[op.qubits[2]]
+
+                elif isinstance(gate, ops.MeasurementGate):
+                    qubits_in_order = op.qubits
+                    ##add the new instance of a key to the numpy array in results dictionary
+                    if (gate.key) in results_dict.keys():
+                        shape = len(qubits_in_order)
+                        current_array = results_dict[gate.key]
+                        new_instance = np.zeros(shape, dtype=np.uint8)
+                        for bits in range(0, len(qubits_in_order)):
+                            new_instance[bits] = values_dict[qubits_in_order[bits]]
+                            results_dict[gate.key] = np.insert(
+                                current_array, len(current_array[0]), new_instance, axis=1
+                            )
+                    else:
+                        ##create the array for the results dictionary
+                        new_array_shape = (repetitions, 1, len(qubits_in_order))
+                        new_array = np.zeros(new_array_shape, dtype=np.uint8)
+                        for reps in range(0, repetitions):
+                            for instances in range(0, 1):
```

Comment:
`range(0, 1)` will just iterate over `0` - use the constant instead of the loop.